app/controllers/storage_controller.rb

```diff
@@ -121,78 +122,57 @@ def show(record = nil)
     @lastaction = "show"
   end
 
+  def specific_buttons(pressed)
+    if button_press_map[pressed]
+      send(button_press_map[pressed])
+    end
 
-  # handle buttons pressed on the button bar
-  def button
-    @edit = session[:edit]                                  # Restore @edit for adv search box
-    params[:display] = @display if %w(all_vms vms hosts).include?(@display) # Were we displaying vms or hosts
+    handle_host_power_buttons(pressed)
 
-    if params[:pressed].starts_with?("vm_", # Handle buttons from sub-items screen
-                                     "miq_template_",
-                                     "guest_",
-                                     "host_")
+    button_press_map.keys.include?(pressed)
+  end
 
-      scanhosts if params[:pressed] == "host_scan"
-      analyze_check_compliance_hosts if params[:pressed] == "host_analyze_check_compliance"
-      check_compliance_hosts if params[:pressed] == "host_check_compliance"
-      refreshhosts   if params[:pressed] == "host_refresh"
-      tag(Host) if params[:pressed] == "host_tag"
-      assign_policies(Host) if params[:pressed] == "host_protect"
-      edit_record  if params[:pressed] == "host_edit"
-      deletehosts if params[:pressed] == "host_delete"
+  def button_press_map
```

Comment:
you are constructing the hash 2x: once to test, then to actually find the symbol

would be better if the hash was a constant.

Maybe you could figure how to make it similar to `generic_x_button`, although the interface is different. But would be nice to unify the interface.